blog/content/post/2020-07-01-dev_and_iommu.md

```diff
@@ -0,0 +1,303 @@
+
+Two relatively simple solutions come to mind:
+
+1. put the relevant tables outside of SLB, or
+2. put the tables inside the SLB and disable DEV before enabling IOMMU.
+
+Both of these ideas leave a time window in which memory holding what will become
+the IOMMU tables is not protected against DMA attacks. If during that window a
+rogue device manages to overwrite the original tables with its own copy, which
+allows that device to have unrestricted access to all RAM, it can basically take
+control of the platform. Even if all measurements were valid we cannot be sure
+that the code was not changed afterwards.
+
+As neither option is safe, we decided to take a deeper look at another
+possibility. We know that IOMMU gets the error when it tries to read the data
+from SLB, so we decided to test how this impact its behaviour.
+
+### IOMMU default state and cache
+
+IOMMU caches the translation tables. If they are changed, the cache must be
+invalidated. This is done by writing a command to the IOMMU command buffer,
+which is one of IOMMU structures in memory, so it should be protected like all
+other IOMMU tables. If IOMMU cannot access that command we have to assume that
+the cache is not invalidated and cannot be trusted.
+
+The test logic is simple: write some known values to memory, order legacy ISA
+DMA engine to overwrite it, read back that memory and compare with initial
+values. The memory range we choose is the first 20 bytes (21 actually, because
+of too safe `memset` function and implicit +1 added to DMA size) of memory,
+which is normally occupied by a real mode Interrupt Vector Table. This region
+was chosen because it must be in the first 1MB of memory (legacy DMA does not
+support higher addresses) and IVT is no longer used, so nothing bad happens if
+it gets corrupted. Initial value was written with `memset(_p(1), 0xcc, 0x20)`,
+and it was later overwritten (or not) with zeros by DMA, as apparently this is
+the value of the idle bus. The memory after the test is dumped twice to make
+sure that DMA engine had enough time to complete the transaction, even though
+printing through UART probably delays execution long enough already.
+
+The following tests were run on apu2, Fam17h does not support (or disables?)
+legacy ISA DMA. They use a variation of what is described under memory to memory
+DMA [here](http://www.osdever.net/tutorials/view/how-to-program-the-dma). Code
+for the final test can be found [here](https://github.com/3mdeb/landing-zone/tree/test_iommu).
+
+As a proof that SLB protection works we took the fact that IOMMU is unable to
+read it, it is not possible to test it with the same ISA DMA engine because LZ
+is loaded into the higher than 1MB addresses. Unless otherwise specified, the
+DMA trial happens between each point below, as well as before and after the
+whole test. As a general rule, we tried to do the simplest possible protection,
+which is "deny all DMA".
+
+##### Test 1 (pass)
```

Comment:
@krystian-hebel I would add when we launch SKINIT in all these steps

Reply:
Added info in paragraph above, it is common for all tests